### PR TITLE
ReOpen PR Support for nameless OKOK scales (Myria MY4836)

### PR DESCRIPTION
**Pull request by oliexdev**
Reopen PR https://github.com/oliexdev/openScale/pull/1081

### CODE REVIEW

**Comment by TRex22** [comment]
So I've been investigating the Bluetooth payload... We could possibly use the manufacturer details or even look for a weight value and only display devices which respond with that data.

Would need to change the code which scans for new devices and might make pairing a little janky

**Comment by oliexdev** [comment]
thanks @TRex22 best think would be to have some manufactur id for filtering

**Comment by TRex22** [comment]
So Im not 100% that all scales have that ID set. I believe the one I have is empty 🤦🏻 
Maybe a simple solution is a check box option in the bluetooth settings (by default off) which enables listing devices with no names. That way People with these devices can use that behaviour and by default those scales + other devices would be hidden

**Comment by oliexdev** [comment]
That will only help more experience users as normally you don't know the Bluetooth address of the scale.
Maybe we can find out how the vendor app does the filtering?

**Comment by Florin9doi** [comment]
The filtering logic is known, it just needs to be integrated with BluetoothSettingsFragment:
https://github.com/oliexdev/openScale/blob/29dd632a8c0b406ce9929478f51ee848aa76846d/android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothOKOK2.java#L165-L169
https://github.com/oliexdev/openScale/blob/29dd632a8c0b406ce9929478f51ee848aa76846d/android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothOKOK2.java#L74-L99

**Comment by oliexdev** [comment]
I created a new branch https://github.com/oliexdev/openScale/tree/MY4836 (sorry don't know how to pushed it into this PR) with a potential solution. I would be happy if somebody could test it as I don't own this kind of nameless scale. The basic idea is to give a nameless scale a fake name which can be used later for identification, see the commit https://github.com/oliexdev/openScale/commit/c17fff3a87ec7576bb84ce2728fbfe4992938cb5.

File: android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothOKOK2.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    private final BluetoothCentralManagerCallback btCallback = new BluetoothCentralManagerCallback() {
+        @Override
+        public void onDiscoveredPeripheral(@NotNull BluetoothPeripheral peripheral, @NotNull ScanResult scanResult) {
+            SparseArray<byte[]> manufacturerSpecificData = scanResult.getScanRecord().getManufacturerSpecificData();
+            int vendorIndex = -1;
+            for (int i = 0; i < manufacturerSpecificData.size(); i++) {
+                int vendorId = manufacturerSpecificData.keyAt(i);
+                if ((vendorId & 0xff) == 0xc0) { // 0x00c0-->0xffc0
```
Comment (oliexdev) [inline android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothOKOK2.java:76]: Is the 0xc0 the identifier for the scale?

Comment (Florin9doi) [reply android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothOKOK2.java:76]: The ID is 0x??c0 (0x00c0, 0x01c0, 0x02c0.... ) where the MSB is used to notify the progress, it starts from 0 and increases until the measurement is stable

**Comment by Florin9doi** [comment]
> I created a new branch https://github.com/oliexdev/openScale/tree/MY4836 (sorry don't know how to pushed it into this PR) with a potential solution. I would be happy if somebody could test it as I don't own this kind of nameless scale. The basic idea is to give a nameless scale a fake name which can be used later for identification, see the commit [c17fff3](https://github.com/oliexdev/openScale/commit/c17fff3a87ec7576bb84ce2728fbfe4992938cb5).

I had to do some changes:
- return if deviceName is still null
- Update deviceView with the new name
- Save the new name in SysPrefs
```diff
diff --git a/android_app/app/src/main/java/com/health/openscale/gui/preferences/BluetoothSettingsFragment.java b/android_app/app/src/main/java/com/health/openscale/gui/preferences/BluetoothSettingsFragment.java
index 7bea7c1c..214ba506 100644
--- a/android_app/app/src/main/java/com/health/openscale/gui/preferences/BluetoothSettingsFragment.java
+++ b/android_app/app/src/main/java/com/health/openscale/gui/preferences/BluetoothSettingsFragment.java
@@ -322,14 +322,19 @@ public class BluetoothSettingsFragment extends Fragment {
             return;
         }
 
-        BluetoothDeviceView deviceView = new BluetoothDeviceView(context);
-        deviceView.setDeviceName(formatDeviceName(bleScanResult.getDevice()));
-
         String deviceName = device.getName();
         if (deviceName == null) {
             deviceName = BluetoothFactory.convertNoNameToDeviceName(bleScanResult.getScanRecord().getManufacturerSpecificData());
+        }
+        if (deviceName == null) {
+            return;
         }
 
+        BluetoothDeviceView deviceView = new BluetoothDeviceView(context);
+        deviceView.setDeviceName(formatDeviceName(deviceName, device.getAddress()));
+        deviceView.setAlias(deviceName);
+
         BluetoothCommunication btDevice = BluetoothFactory.createDeviceDriver(context, deviceName);
         if (btDevice != null) {
             Timber.d("Found supported device %s (driver: %s)",
@@ -398,6 +403,7 @@ public class BluetoothSettingsFragment extends Fragment {
         private TextView deviceName;
         private ImageView deviceIcon;
         private String deviceAddress;
+        private String deviceAlias;
 
         public BluetoothDeviceView(Context context) {
             super(context);
@@ -439,6 +445,14 @@ public class BluetoothSettingsFragment extends Fragment {
             deviceName.setText(name);
         }
 
+        public void setAlias(String alias) {
+            deviceAlias = alias;
+        }
+
+        public String getAlias() {
+            return deviceAlias;
+        }
+
         public void setSummaryText(String text) {
             SpannableStringBuilder stringBuilder = new SpannableStringBuilder(new String());
 
@@ -485,10 +499,10 @@ public class BluetoothSettingsFragment extends Fragment {
 
             prefs.edit()
                     .putString(PREFERENCE_KEY_BLUETOOTH_HW_ADDRESS, device.getAddress())
-                    .putString(PREFERENCE_KEY_BLUETOOTH_DEVICE_NAME, device.getName())
+                    .putString(PREFERENCE_KEY_BLUETOOTH_DEVICE_NAME, getAlias())
                     .apply();
 
-            Timber.d("Saved Bluetooth device " + device.getName() + " with address " + device.getAddress());
+            Timber.d("Saved Bluetooth device " + getAlias() + " with address " + device.getAddress());
 
             stopBluetoothDiscovery();
```

**Comment by oliexdev** [comment]
Thanks @Florin9doi I applied the patch with https://github.com/oliexdev/openScale/commit/b232995d47e83cc7ad61bfcddd60d5955afaa9a2

Does this version work now with the nameless OkOk scale?

**Comment by TRex22** [comment]
If you'd like I can generate an APK on my side and test? ... probably only over the weekend when I find some free time.

**Comment by Florin9doi** [comment]
> Thanks @Florin9doi I applied the patch with [b232995](https://github.com/oliexdev/openScale/commit/b232995d47e83cc7ad61bfcddd60d5955afaa9a2)
> 
> Does this version work now with the nameless OkOk scale?

It works, this is how it looks : 
image

**Comment by Florin9doi** [comment]
These 2 comments can be removed: https://github.com/oliexdev/openScale/blob/b232995d47e83cc7ad61bfcddd60d5955afaa9a2/android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothOKOK2.java#L175-L176
I tested on a S24U with Android 14 and [doesn't work](https://android.googlesource.com/platform/packages/modules/Bluetooth/+/6310469dfbe98cccec0e214294b1f3e47b22e99d/system/gd/hci/le_scanning_manager.cc#978).

**Comment by oliexdev** [comment]
thanks for testing, I will merge it

**Comment by Florin9doi** [comment]
Are you sure you merged the correct branch? I can't find in master the latest fixes from https://github.com/oliexdev/openScale/tree/MY4836

**Comment by oliexdev** [comment]
you are absolutely right, I hope I pushed the right one now, could you test it maybe again?

**Comment by Florin9doi** [comment]
I tested the latest dev build (https://github.com/oliexdev/openScale/releases/tag/dev-build) and it is ok. 
Thank you

**Comment by oliexdev** [comment]
Thanks for testing and your great PR

**Pull request merged**